### DISCUSSION TITLE
Why does "The request was canceled before receiving a response" happen?

### DISCUSSION THREAD

**Discussion by GingerMoon** [dotnet/yarp#1944] [Q&A] [answered]
I found a similiar case https://github.com/microsoft/reverse-proxy/discussions/710.
But I don't think our case is because of end users refreshes UI pages. 
This is because the time window (from the reverse proxy proxying the request, to the target app returns the response to the reverse proxy) is only 1 second. However, the reverse proxy still canceled the request before receiving a response.
I am wondering if there are any other reasons why this happen?

We are using reverse proxy (HttpForwarder) in our projects. 
The timeout is 100 seconds.
Below are our logs:
```
01/12/22 16:58:43.459   Proxying to ..... HTTP/2 RequestVersionOrLower no-streaming -> reverse proxy log
01/12/22 16:58:43.953   RequestCanceled: The request was canceled before receiving a response. -> reverse proxy log
01/12/22 16:58:44.415   target app returns 200 to the reverse proxy  -> target app log
```

By the way, I also find that sometime (not very often), the target app is invoked more than once, although the reverse proxy proxies the request to the target app only once. 
Did anyone else report such case?

**Comment by MihaZupan** [comment, marked as answer]
> But I don't think our case is because of end users refreshes UI pages.
This is because the time window (from the reverse proxy proxying the request, to the target app returns the response to the reverse proxy) is only 1 second.

The client disconnecting is the only condition where you'll see `RequestCanceled`.
It doesn't really matter how long the destination takes to reply, 1 second is a long window. You should be able to reproduce that by just hitting F5 in the browser quickly a few times.

> By the way, I also find that sometime (not very often), the target app is invoked more than once, although the reverse proxy proxies the request to the target app only once.
Did anyone else report such case?

Not that I'm aware of. The proxy doesn't do any retries on its own. Did you see the app receive multiple requests from the proxy?

**Reply by GingerMoon** [reply]
Thanks MihaZupan, thanks for the reply.
Yes, according to our log, the target app receives multiple requests. 
But it doesn't happen very often.

**Reply by MihaZupan** [reply]
And these requests are all the same, and you can confirm that there's more of them than the number of requests that hit YARP? That is surprising.
Is the destination receiving HTTP/1 or 2 requests in this case?
If it's possible to reproduce without TLS, would you be able to collect a network trace when that happens?
